arrabiata/src/witness.rs

```diff
@@ -0,0 +1,1098 @@
+use kimchi::circuits::domains::EvaluationDomains;
+use log::{debug, info};
+use num_bigint::{BigInt, BigUint};
+use num_integer::Integer;
+use o1_utils::field_helpers::FieldHelpers;
+use poly_commitment::{commitment::CommitmentCurve, srs::SRS, PolyComm, SRS as _};
+use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
+use std::time::Instant;
+
+use crate::{
+    columns::{Column, Gadget},
+    interpreter::{Instruction, InterpreterEnv, Side},
+    poseidon_3_60_0_5_5_fp, poseidon_3_60_0_5_5_fq, BIT_DECOMPOSITION_NUMBER_OF_CHUNKS,
+    MAXIMUM_FIELD_SIZE_IN_BITS, NUMBER_OF_COLUMNS, NUMBER_OF_PUBLIC_INPUTS, NUMBER_OF_SELECTORS,
+    NUMBER_OF_VALUES_TO_ABSORB_PUBLIC_IO, POSEIDON_ALPHA, POSEIDON_ROUNDS_FULL,
+    POSEIDON_STATE_SIZE,
+};
+
+pub const IVC_STARTING_INSTRUCTION: Instruction = Instruction::Poseidon(0);
+
+/// An environment that can be shared between IVC instances.
+///
+/// It contains all the accumulators that can be picked for a given fold
+/// instance k, including the sponges.
+///
+/// The environment is run over big integers to avoid performing
+/// reduction at all step. Instead the user implementing the interpreter can
+/// reduce in the corresponding field when they want.
+pub struct Env<
+    Fp: PrimeField,
+    Fq: PrimeField,
+    E1: AffineCurve<ScalarField = Fp, BaseField = Fq>,
+    E2: AffineCurve<ScalarField = Fq, BaseField = Fp>,
+> {
+    // ----------------
+    // Setup related (domains + SRS)
+    /// Domain for Fp
+    pub domain_fp: EvaluationDomains<Fp>,
+
+    /// Domain for Fq
+    pub domain_fq: EvaluationDomains<Fq>,
+
+    /// SRS for the first curve
+    pub srs_e1: SRS<E1>,
+
+    /// SRS for the second curve
+    pub srs_e2: SRS<E2>,
+    // ----------------
+
+    // ----------------
+    // Information related to the IVC, which will be used by the prover/verifier
+    // at the end of the whole execution
+    // FIXME: use a blinded comm and also fold the blinder
+    pub ivc_accumulator_e1: Vec<PolyComm<E1>>,
+
+    // FIXME: use a blinded comm and also fold the blinder
+    pub ivc_accumulator_e2: Vec<PolyComm<E2>>,
+
+    /// Commitments to the previous instances
+    pub previous_commitments_e1: Vec<PolyComm<E1>>,
+    pub previous_commitments_e2: Vec<PolyComm<E2>>,
+    // ----------------
+
+    // ----------------
+    // Data only used by the interpreter while building the witness over time
+    /// The index of the latest allocated variable in the circuit.
+    /// It is used to allocate new variables without having to keep track of the
+    /// position.
+    pub idx_var: usize,
+
+    /// The index of the latest allocated public inputs in the circuit.
+    /// It is used to allocate new public inputs without having to keep track of
+    /// the position.
+    pub idx_var_pi: usize,
+
+    /// Current processing row. Used to build the witness.
+    pub current_row: usize,
+
+    /// State of the current row in the execution trace
+    pub state: [BigInt; NUMBER_OF_COLUMNS],
```

Comment:
We will need two vectors like this when we will have access to the next row ?

Reply:
Yes, it is the case, see new PR.